### ISSUE TITLE
Support exclusive and half-open intervals in BETWEEN expressions

### ISSUE THREAD

**Issue by nwoolmer** [questdb/questdb#4746]
### Is your feature request related to a problem?

QuestDB supports BETWEEN expressions:

```sql
WHERE timestamp BETWEEN x AND y
```

These expressions serve as a convenient syntax for representing inclusive intervals. This is syntax sugar for:

```sql
WHERE timestamp >= x AND timestamp <= y
```

For many applications, an inclusive interval is fine. However, there are other types of intervals that can be useful to represent: inclusive, exclusive, left-open, right-open.

We could extend the BETWEEN syntax to support these intervals.

This can also serve as a small performance optimisation, since both conditions will be calculated in a single function call.

### Describe the solution you'd like.

New syntax to cover the other three main cases of intervals:

- [x] Inclusive: `WHERE timestamp >= x AND timestamp <= y`
    - Current syntax: `WHERE timestamp BETWEEN x AND y`
    - Extension: `WHERE timestamp BETWEEN INCLUSIVE x AND y` 
    - (but default remains as-is, not necessary to specify inclusive)
    
- [ ] Exclusive: `WHERE timestamp > x and timestamp < y`
    - Extension: `WHERE timestamp BETWEEN EXCLUSIVE x AND y` 
    
- [ ] Left-open: `WHERE timestamp > x AND timestamp <= y`
    - Extension: `WHERE timestamp BETWEEN LEFT OPEN x AND y`
    
- [ ] Right-open: `WHERE timestamp >= x AND timestamp < y`
    - Extension: `WHERE timestamp BETWEEN RIGHT OPEN x AND y`
    

The extra terms could also be placed at the end i.e:

`WHERE timestamp BETWEEN x AND y EXCLUSIVE`

### Describe alternatives you've considered.

Using `<`, `<=`, `>`, `>=` operators.

### Full Name:

Nick Woolmer

### Affiliation:

QuestDB

**Comment by Shikharishere** [comment]
Hey @nwoolmer I would like to take up this issue. Can you suggest me a starting point.

**Comment by Shikharishere** [comment]
Hey, @nwoolmer let me clear my understanding. You meant that the BETWEEN clause should also support these ranges like inclusive, exclusive, left-open, and right-open. And thanks for the suggestions.

**Comment by Shikharishere** [comment]
Hey, @nwoolmer I have setup Questdb and I have learned some of the syntax (that is rather similar to SQL) and now I will start developing the extended support for the BETWEEN clause.

**Comment by Shikharishere** [comment]
> Hi @Shikharishere ,
> 
> By all means, thank you for your interest.
> 
> Firstly, I think placing the new keywords at the end of the clause reads better i.e
> 
> `BETWEEN x AND y EXCLUSIVE`
> 
> What do you think?
> 
> Here are some suggestions of how you could implement this:
> 
> * Add a new parameter to QueryModel which defines the type of the BETWEEN expression. See how other constants are defined already as a guide (constants instead of enums)
> * Update SqlParser to handle the new syntax and set those fields
> * Create new factories like `BetweenTimestampFunctionFactory`, and if necessary, update `SqlCodeGenerator` to handle the new cases. I would suggest looking at `TimestampFloorFunctionFactory` for inspiration.
> * Add appropriate happy path, and error path tests for each of these components
> * (optional) Update the corresponding documentation for BETWEEN in the docs repo (which we recently open sourced)
> 
> The current behaviour of BETWEEN should remain unchanged, and be equivalent to:
> 
> `BETWEEN x AND y INCLUSIVE`
> 
> Feel free to ask for feedback and help on here, start a thread on QuestDB community, or on slack.
> 
> You are also welcome to open a draft PR whenever you would like.
> 
> Thanks!

@nwoolmer I do agree with you placing the new keywords at the end of the query.

**Comment by Shikharishere** [comment]
`

    private ExpressionNode betweenStart;
    private ExpressionNode betweenEnd;
 
    public ExpressionNode getBetweenStart() {
        return betweenStart;
    }

    public ExpressionNode getBetweenEnd() {
        return betweenEnd;
    }

    public void setBetweenStart(ExpressionNode betweenStart) {
        this.betweenStart = betweenStart;
    }

    public void setBetweenEnd(ExpressionNode betweenEnd) {
        this.betweenEnd = betweenEnd;
    }
`

@nwoolmer This is how I am going with step 1 you suggested. Do you think I also need to add changes to clear, equals and hashcode methods? I think I have to ...

**Comment by Shikharishere** [comment]
Yeah, I found the expression. This is where I should add the INCLUSIVE and EXCLUSIVE keywords, I'm still confused about the code flow. In my understanding, the query model class builds the query and the query parser parses it.
 **Could you be more verbose about how can I add a new clause to the query model, so that it can build the query with that clause?** 
I also think we may need to add more methods to windowColumn, especially for these two classes. I also found out that the SQL keywords class doesn't have a method like inInclusiveKeyword so that also needs to be added.

**Comment by Shikharishere** [comment]
Constants : 
```
    public static final int INCLUSIVE_BETWEEN = 1;
    public static final int EXCLUSIVE_BETWEEN = 2;
    public static final int RIGHT_OPEN_BETWEEN = 3;
    public static final int LEFT_OPEN_BETWEEN = 4;
```
The field : 
`private int betweenType = INCLUSIVE_BETWEEN;`

Also created the getter and setters for this, created the `isInclusiveKeyword` method as well. 
As you suggest, there's nothing to be done in QueryMode class? 
Or I need to do something like this sink.put( " INCLUSIVE ")?

**Comment by Shikharishere** [comment]
Yeah, I initialized betweenType field with ``INCLUSIVE_BETWEEN``.
I also update the methods like hashCode, clear and equals as well. 
Okay, I will open a draft pr.

**Comment by Shikharishere** [comment]
Hey, did you check out the pull request?
